### PR TITLE
[Fixes #852] TestHost: OnStarting and OnCompleted callbacks of response are not being awaited

### LINKED ISSUE #852: TestHost: OnStarting and OnCompleted callbacks of response are not being awaited
https://github.com/aspnet/Hosting/blob/dev/src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs#L42

Since they are not being awaited, the code within these call backs get executed at random times causing unexpected failures (ex: In the delegate accessing session on HttpContext would fail if the code gets executed way late by which the session middleware has removed the session from the context).

From @Tratcher : These call backs' signature used to be synchronous earlier and was later changed to async, but we missed changing the call sites to consider the new async behavior.

cc @rynowak

### PR DESCRIPTION
**Pull request by kichalla**
@Tratcher

### CODE REVIEW

File: src/Microsoft.AspNetCore.TestHost/ClientHandler.cs [hunk 1]
```diff
@@ -187,27 +187,26 @@ internal void AbortRequest()
...
             }
 
-            internal void ReturnResponseMessage()
+            internal async Task ReturnResponseMessageAsync()
             {
                 if (!_responseTcs.Task.IsCompleted)
                 {
-                    var response = GenerateResponse();
-                    // Dispatch, as TrySetResult will synchronously execute the waiters callback and block our Write.
-                    Task.Factory.StartNew(() => _responseTcs.TrySetResult(response));
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.TestHost/ClientHandler.cs]: I did not see any issues running tests with this change.

Comment (Tratcher) [inline src/Microsoft.AspNetCore.TestHost/ClientHandler.cs]: Please revert this, calling SetResult inline can cause threading issues and hangs

File: src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs [hunk 2]
```diff
@@ -36,33 +36,42 @@ public ResponseFeature()
...
+            var prior = _responseCompletedAsync;
+            _responseCompletedAsync = async () =>
             {
-                callback(state);
-                prior();
+                try
+                {
+                    await callback(state);
+                }
+                finally { }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs]: this try/finally doesn't do anything, your exception gets rethrown after the finally. `await prior();` should be in the finally block

Comment (Tratcher) [inline src/Microsoft.AspNetCore.TestHost/ResponseFeature.cs]: In a normal server we'd catch all the exceptions and log them, but here we stand a chance of surfacing them to the client. See if you can make one throw all the way back to the client.

File: src/Microsoft.AspNetCore.TestHost/ClientHandler.cs [hunk 3]
```diff
@@ -187,27 +187,33 @@ internal void AbortRequest()
...
+                if (!Context.HttpContext.Response.HasStarted)
                 {
-                    var response = GenerateResponse();
-                    // Dispatch, as TrySetResult will synchronously execute the waiters callback and block our Write.
-                    Task.Factory.StartNew(() => _responseTcs.TrySetResult(response));
+                    if (!_responseTcs.Task.IsCompleted)
+                    {
+                        var response = await GenerateResponseAsync();
+                        // Dispatch, as TrySetResult will synchronously execute the waiters callback and block our Write.
+                        var setResult = Task.Factory.StartNew(() => _responseTcs.TrySetResult(response));
```
Comment (campersau) [inline src/Microsoft.AspNetCore.TestHost/ClientHandler.cs]: Looks like `setResult` is unused?

Comment (Tratcher) [inline src/Microsoft.AspNetCore.TestHost/ClientHandler.cs]: It's set to avoid a compiler warning about not awaiting the task.

**Comment by kichalla** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.TestHost/ClientHandler.cs [hunk 4]
```diff
@@ -187,27 +187,33 @@ internal void AbortRequest()
...
-                if (!_responseTcs.Task.IsCompleted)
+                // Check if the response has already started because the TrySetResult below could happen a bit late
+                // (as it happens on a different thread) by which point the CompleteResponseAsync could run and calls this
+                // method again.
+                if (!Context.HttpContext.Response.HasStarted)
                 {
-                    var response = GenerateResponse();
-                    // Dispatch, as TrySetResult will synchronously execute the waiters callback and block our Write.
-                    Task.Factory.StartNew(() => _responseTcs.TrySetResult(response));
+                    if (!_responseTcs.Task.IsCompleted)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.TestHost/ClientHandler.cs]: You shouldn't need to check both HasStarted and IsCompleted. HasStarted should be sufficient.

Comment (kichalla) [inline src/Microsoft.AspNetCore.TestHost/ClientHandler.cs]: Sounds good. :up: :date:

**Pull request merged**